src/common/rabbitmq/rabbitmq.service.ts

```diff
@@ -0,0 +1,116 @@
+import amqp, { Connection, Channel } from "amqplib/callback_api";
+import { Injectable, Logger, OnModuleDestroy } from "@nestjs/common";
+import { ConfigService } from "@nestjs/config";
+
+/**
+ * Service for publishing messages to a RabbitMQ queue.
+ */
+@Injectable()
+export class RabbitMQService implements OnModuleDestroy {
+  private connectionOptions: amqp.Options.Connect;
+  private connection: Connection;
+  private channel: Channel;
+
+  constructor(private configService: ConfigService) {
+    Logger.log("Initializing RabbitMQService.", "RabbitMQService");
+
+    const rabbitMqEnabled =
+      this.configService.get<string>("rabbitMq.enabled") === "yes";
+
+    if (rabbitMqEnabled) {
+      const hostname = this.configService.get<string>("rabbitMq.hostname");
+      const port = this.configService.get<number>("rabbitMq.port");
+      const username = this.configService.get<string>("rabbitMq.username");
+      const password = this.configService.get<string>("rabbitMq.password");
+
+      if (!hostname || !port || !username || !password) {
+        Logger.error(
+          "RabbitMQ is enabled but missing one or more config variables.",
+          "RabbitMQService",
+        );
+      } else {
+        this.connectionOptions = {
+          protocol: "amqp",
+          hostname: hostname,
+          port: port,
+          username: username,
+          password: password,
+        };
+
+        amqp.connect(
+          this.connectionOptions,
+          (connectionError: Error, connection: Connection) => {
+            if (connectionError) {
+              Logger.error(
+                "Connection error in RabbitMQService: " +
+                  JSON.stringify(connectionError.message),
+                "RabbitMQService",
+              );
+              return;
+            }
+            this.connection = connection;
+
+            this.connection.createChannel(
+              (channelError: Error, channel: Channel) => {
+                if (channelError) {
+                  Logger.error(
+                    "Channel error in RabbitMQService: " +
+                      JSON.stringify(channelError.message),
+                    "RabbitMQService",
+                  );
+                  return;
+                }
+                this.channel = channel;
+                Logger.log(this.channel);
+              },
+            );
+          },
+        );
+      }
+    } else {
+      Logger.error("RabbitMQ is not enabled.", "RabbitMQService");
```

Comment:
What's the expected behavior if they don't have RabbitMQ enabled? I think the service gets instantiated either way. Maybe the constructor should succeed (without logging) but there should be an error if connect() is called or if a rabbitmqaction is created.